### ISSUE TITLE
CMake generates a solution with Multi Byte Character Set

### ISSUE THREAD

**Issue by degski** [bluescarni/mppp#177]
The generated solution uses the multi-byte character set. Although this seems easier, it generates inefficient code as any interaction with the OS is converted to and from UNICODE, as windows is strictly UNICODE internally (their messy flavour of it). Not sure whether this is extremely important, but I thought I'd just mention it.

In the meanwhile, I've sorted out (with Brian) some issues related to MPIR and have now broadwell_avx native libs [and MPFR, also Brian's]. Now I've sorted the whole thing out, the issues I referred to in the

**Comment by bluescarni** [comment]
I've done some reading and it seems like it's possible to tell CMake to create a Unicode-based solution with a couple of definitions. Are there particular caveats, incompatibilities, etc. I should know about before trying to change this setting?

**Comment by degski** [comment]
I've just opened a new default console project solution (it's a MS supplied project template) and, the only thing I can see is that `/D "_UNICODE" /D "UNICODE"` is added  to the compile line/commands.

To be precise, this is the while line that gets thrown at a file (static exe):

    /permissive- /Yu"pch.h" /GS /GL /W3 /Gy /Zc:wchar_t /Zi /Gm- /O2 /sdl /Fd"x64\Release\vc141.pdb" /Zc:inline /fp:precise /D "NDEBUG" /D "_CONSOLE" /D "_UNICODE" /D "UNICODE" /errorReport:prompt /WX- /Zc:forScope /Gd /Oi /MT /FC /Fa"x64\Release\" /EHsc /nologo /Fo"x64\Release\" /Fp"x64\Release\ConsoleApplication1.pch" /diagnostics:classic

**Comment by bluescarni** [comment]
Ok, I'll give it a go then :) This applies to both MSVC (``cl.exe``) and ``clang-cl`` right?

**Comment by bluescarni** [comment]
I have added the defines in #179. Do you reckon I should have an option in the build system to specify whether to enable unicode support or not? I am a bit afraid of possible binary incompatibilites down the line...

**Comment by degski** [comment]
@bluescarni Hi, I just after some of your changes did a pull and rebuild. The generated solution still is `MBCS`. 

> Since they can be set by the user when compiling mp++ (e.g., via CXXFLAGS), I don't think it's necessary to explicitly add an option in the build system for them.

You're thinking make etc [nix-thinking]. Using VS, one does not explicitly set CXXFLAGS (nor CFLAGS). It's hard to imagine that `CMake` would make the project default to `MBCS` without being explicitly told to do so [do you see what I'm implying?].

**Comment by bluescarni** [comment]
@degski which commit did you rebuild? If you rebuilt with the latest master, could you try building from this commit

https://github.com/bluescarni/mppp/pull/179/commits/4543280e57c6565eadc8a58db68e8adf043dd340

and tell me if it produces a unicode solution?

**Comment by degski** [comment]
@bluescarni Yes, it [that commit] does generate a unicode-solution, and `/D "UNICODE" /D "_UNICODE"` get added as well.

**Comment by bluescarni** [comment]
Ok, I have a new PR up at #183. I added a build option called ``MPPP_MSVC_UNICODE``, which you'll have to set to ``YES`` in order to get a Unicode solution.

> It's hard to imagine that CMake would make the project default to MBCS without being explicitly told to do so [do you see what I'm implying?].

I have never been aware of any charset-related option in CMake before you opened this PR. Actually, to be precise, even here I am only setting some flags/definitions for the compiler - so it looks like either CMake is parsing the compilation flags and deciding that a Unicode solution needs to be created, or perhaps MSVC is doing it. I have no idea which is which.

My guess is that the CMake devs are just being conservative. In any case, setting those definitions effectively overrides a CMake default, hence the decision of making ``MPPP_MSVC_UNICODE`` an opt-in setting, rather than the default choice.

**Comment by degski** [comment]
@bluescarni On Windows `UNICODE` is not an opt-in, iff you want to stand any chance you're software will work correctly, you will have to use `UNICODE` and all the [lovely] stuff it brings with it ;-) .

I'm tired arguing now though and there is a working option, so I'll drop my opposition and just specify whatever flag required [I presume you will document those things].

Thanks for your efforts, well appreciated.
